filebeat/input/v2/statestore/store.go

```diff
@@ -0,0 +1,176 @@
+// If one of the go-routines closes the store, we ensure that the shared resources
+// are kept alive until all go-routines have given up access.
+type store struct {
+	refCount concert.RefCount
+
+	parent *Connector
+	name   string
+
+	persistentStore *registry.Store
+	resourcesMux    sync.Mutex
+	resources       table
+}
+
+// storeSession keeps track of the lifetime of a Store instance.
+// In flight resource update operations do extend the lifetime of
+// a Store, even if the store has been closed by a go-routine.
+//
+// A session will shutdown when the Store is closed and all pending
+// update operations have been persisted.
+type storeSession struct {
+	refCount concert.RefCount
+	store    *store
+}
+
+func newSession(store *store) *storeSession {
+	session := &storeSession{store: store}
+	session.refCount.Action = func(_ error) { session.Close() }
+	return session
+}
+
+func (s *storeSession) Close() {
+	connector := s.store.parent
+	connector.releaseStore(s.store)
+	s.store = nil
+}
+
+func (s *storeSession) Retain()       { s.refCount.Retain() }
+func (s *storeSession) Release() bool { return s.refCount.Release() }
+
+// newStore creates a new state Store with an active session.
+func newStore(store *store) *Store {
+	invariant(store != nil, "missing a persistent store")
+
+	return &Store{
+		active:  atomic.MakeBool(true),
+		store:   store,
+		session: newSession(store),
+	}
+}
+
+// Close deactivates the store and waits for all resources to be released.
+// Resources can not be accessed anymore, but in progress resource updates are
+// still active, until they are eventually ACKed.  The underlying persistent
+// store will be finally closed once all pending updates have been written to
+// the persistent store.
+func (s *Store) Close() {
+	s.active.Store(false)
+	s.session.Release()
+}
+
+// Access an unlocked resource. This creates a handle to a resource that may
+// not yet exist in the persistent registry.
+func (s *Store) Access(key ResourceKey) *Resource {
+	return newResource(s, key)
+}
+
+func (s *Store) findOrCreate(key ResourceKey, lm lockMode) (res *resourceEntry) {
+	if !s.active.Load() {
+		return nil
+	}
+
+	withLockMode(&s.resourcesMux, lm, func() error {
+		if res = s.resources.Find(key); res == nil {
+			res = s.resources.Create(key)
+		}
+		return nil
+	})
+	return res
+}
+
+// find returns the in memory resource entry, if the key is known to the store.
+// find returns nil if the resource is unknown so far.
+func (s *Store) find(key ResourceKey, lm lockMode) (res *resourceEntry) {
+	if !s.active.Load() {
+		return nil
+	}
+	return s.store.find(key, lm)
+}
+
+// create adds a new entry to the in-memory table and returns the pointer to the entry.
+// It fails if the store has been deactivated and returns 'null'.
+func (s *Store) create(key ResourceKey, lm lockMode) (res *resourceEntry) {
+	if !s.active.Load() {
+		return nil
+	}
+	return s.store.create(key, lm)
+}
+
+func (s *store) close() error {
```

Comment:
Shouldn't `s.parent` be set to `nil` when the `store` is closed?

Reply:
It doesn't matter. This function is called internally after all refcounts are 0 and the shared store is removed from the connector. The shared store doesn't really need the parent and name fields. I will move them to the session.